catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/QueryOperations.java

```diff
@@ -161,115 +169,30 @@ QueryResponse query(QueryRequest queryRequest, FederationStrategy strategy,
-                sourceIds.remove(getId());
-                sourceIds.remove(null);
-                sourceIds.remove("");
-            }
-
-            // See if we still have sources to look up by name
-            if (!sourceIds.isEmpty()) {
-                for (String id : sourceIds) {
-                    LOGGER.debug("Looking up source ID = {}", id);
-                    sourceFound = false;
-                    if (frameworkProperties.getFederatedSources()
-                            .containsKey(id)) {
-                        sourceFound = true;
-                        if (frameworkProperties.getFederatedSources()
-                                .get(id)
-                                .isAvailable()
-                                && canAccessSource(frameworkProperties.getFederatedSources()
-                                .get(id), queryRequest)) {
-                            sourcesToQuery.add(frameworkProperties.getFederatedSources()
-                                    .get(id));
-                        } else {
-                            exceptions.add(createUnavailableProcessingDetails(frameworkProperties.getFederatedSources()
-                                    .get(id)));
-                        }
-                    }
-
-                    if (!sourceFound) {
-                        exceptions.add(new ProcessingDetailsImpl(id,
-                                new SourceUnavailableException("Source id is not found")));
-                    }
-                }
-            }
-        } else {
-            // default to local sources
-            addConnectedSources = connectedSourcesExist();
-            addCatalogProvider = hasCatalogProvider(fanoutEnabled);
-        }
-
-        if (addConnectedSources) {
-            // add Connected Sources
-            for (ConnectedSource source : frameworkProperties.getConnectedSources()) {
-                if (sourceOperations.isSourceAvailable(source)) {
-                    sourcesToQuery.add(source);
-                } else {
-                    LOGGER.debug("Connected Source {} is unavailable and will not be queried.",
-                            source.getId());
-                }
-            }
-        }
 
-        if (addCatalogProvider) {
-            if (sourceOperations.isSourceAvailable(sourceOperations.getCatalog())) {
-                sourcesToQuery.add(sourceOperations.getCatalog());
-            } else {
-                exceptions.add(createUnavailableProcessingDetails(sourceOperations.getCatalog()));
-            }
-        }
+        QuerySources querySources = new QuerySources(this, frameworkProperties).initializeSources(
+                queryRequest,
+                sourceIds,
+                fanoutEnabled)
+                .addConnectedSources(this, frameworkProperties)
+                .addCatalogProvider(this);
 
-        if (sourcesToQuery.isEmpty()) {
+        if (querySources.sourcesToQuery.isEmpty()) {
```

Comment:
Should we have a `isEmpty()` on `QuerySources` instead of having to call it on `sourcesToQuery`?